lib/upgrade/etcd/conversionv1v3/workloadendpoint.go

```diff
@@ -0,0 +1,276 @@
+			ipv4Nets = append(ipv4Nets, n)
+		} else {
+			ipv6Nets = append(ipv6Nets, n)
+		}
+	}
+
+	ipv4NAT := []model.IPNAT{}
+	ipv6NAT := []model.IPNAT{}
+	for _, n := range ah.Spec.IPNATs {
+		nat := model.IPNAT{IntIP: n.InternalIP, ExtIP: n.ExternalIP}
+		if n.InternalIP.Version() == 4 {
+			ipv4NAT = append(ipv4NAT, nat)
+		} else {
+			ipv6NAT = append(ipv6NAT, nat)
+		}
+	}
+
+	var ports []model.EndpointPort
+	for _, port := range ah.Spec.Ports {
+		ports = append(ports, model.EndpointPort{
+			Name:     port.Name,
+			Protocol: port.Protocol,
+			Port:     port.Port,
+		})
+	}
+
+	d := model.KVPair{
+		Key: k,
+		Value: &model.WorkloadEndpoint{
+			Labels:           ah.Metadata.Labels,
+			ActiveInstanceID: ah.Metadata.ActiveInstanceID,
+			State:            "active",
+			Name:             ah.Spec.InterfaceName,
+			Mac:              ah.Spec.MAC,
+			ProfileIDs:       ah.Spec.Profiles,
+			IPv4Nets:         ipv4Nets,
+			IPv6Nets:         ipv6Nets,
+			IPv4NAT:          ipv4NAT,
+			IPv6NAT:          ipv6NAT,
+			IPv4Gateway:      ah.Spec.IPv4Gateway,
+			IPv6Gateway:      ah.Spec.IPv6Gateway,
+			Ports:            ports,
+		},
+		Revision: ah.Metadata.Revision,
+	}
+
+	log.Debugf("Converted: %+v\n To: %+v", ah, d)
+
+	return &d, nil
+}
+
+// BackendV1ToAPIV3 converts v1 WorkloadEndpoint KVPair to v3 API.
+func (_ WorkloadEndpoint) BackendV1ToAPIV3(kvp *model.KVPair) (Resource, error) {
+	wepKey, ok := kvp.Key.(model.WorkloadEndpointKey)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource key")
+	}
+	wepValue, ok := kvp.Value.(model.WorkloadEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource Value")
+	}
+
+	labels := convertLabels(wepValue.Labels)
+	namespace := "default"
+
+	var pod string
+	var container string
+	var workload string
+
+	// Populate our values based on the orchestrator.
+	switch wepKey.OrchestratorID {
+	case "k8s":
+		namespace = labels["projectcalico.org/namespace"]
+		workload = convertName(wepKey.WorkloadID)
+		pod = getPodName(workload, namespace)
+	case "cni":
+		container = wepValue.ActiveInstanceID
+	case "libnetwork":
+		workload = "libnetwork"
+	default:
+		workload = convertName(wepKey.WorkloadID)
+	}
+
+	ipNets := convertIPNetworks(wepValue.IPv4Nets)
+	ipNets = append(ipNets, convertIPNetworks(wepValue.IPv6Nets)...)
+
+	ipNats := convertIPNATs(wepValue.IPv4NAT)
+	ipNats = append(ipNats, convertIPNATs(wepValue.IPv6NAT)...)
+
+	wep := apiv3.WorkloadEndpoint{
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: apiv3.WorkloadEndpointSpec{
+			Orchestrator:  convertName(wepKey.OrchestratorID),
+			Workload:      workload,
+			Node:          convertName(wepKey.Hostname),
+			Pod:           pod,
+			ContainerID:   container,
+			Endpoint:      convertName(wepKey.EndpointID),
+			IPNetworks:    ipNets,
+			IPNATs:        ipNats,
+			Profiles:      convertProfiles(wepValue.ProfileIDs),
```

Comment:
You should share this with @ozdanborne - he need the `convertProfiles` function for host endpoints as well.